main/src/ca/uwaterloo/flix/language/phase/Kinder.scala

```diff
@@ -0,0 +1,1155 @@
+
+      val tparamsVal = Validation.traverse(tparams0.tparams)(visitTypeParam(_, kenv))
+      val tpeVal = visitType(tpe0, KindMatch.Wild, kenv, root)
+
+      mapN(tparamsVal, tpeVal) {
+        case (_, _) => ()
+      }
+  }
+
+  /**
+    * Performs kinding on the given enum case under the given kind environment.
+    */
+  private def visitCase(caze0: ResolvedAst.Case, kenv: KindEnv, root: ResolvedAst.Root)(implicit flix: Flix): Validation[KindedAst.Case, KindError] = caze0 match {
+    case ResolvedAst.Case(enum, tag, tpeDeprecated0, sc0) =>
+      for {
+        tpeDeprecated <- visitType(tpeDeprecated0, KindMatch.subKindOf(Kind.Star), kenv, root)
+        sc <- visitScheme(sc0, kenv, root)
+      } yield KindedAst.Case(enum, tag, tpeDeprecated, sc)
+  }
+
+  /**
+    * Performs kinding on the given type class.
+    */
+  private def visitClass(clazz: ResolvedAst.Class, root: ResolvedAst.Root)(implicit flix: Flix): Validation[KindedAst.Class, KindError] = clazz match {
+    case ResolvedAst.Class(doc, mod, sym, tparam0, superClasses0, sigs0, laws0, loc) =>
+      val kenv = getKindEnvFromTypeParamDefaultStar(tparam0)
+
+      val tparamVal = visitTypeParam(tparam0, kenv)
+      val superClassesVal = Validation.traverse(superClasses0)(visitTypeConstraint(_, kenv, root))
+      val sigsVal = Validation.traverse(sigs0) {
+        case (sigSym, sig0) => visitSig(sig0, kenv, root).map(sig => sigSym -> sig)
+      }
+      val lawsVal = traverse(laws0)(visitDef(_, kenv, root))
+
+      mapN(tparamVal, superClassesVal, sigsVal, lawsVal) {
+        case (tparam, superClasses, sigs, laws) => KindedAst.Class(doc, mod, sym, tparam, superClasses, sigs.toMap, laws, loc)
+      }
+  }
+
+  /**
+    * Performs kinding on the given instance.
+    */
+  private def visitInstance(inst: ResolvedAst.Instance, root: ResolvedAst.Root)(implicit flix: Flix): Validation[KindedAst.Instance, KindError] = inst match {
+    case ResolvedAst.Instance(doc, mod, sym, tpe0, tconstrs0, defs0, ns, loc) =>
+      val kind = getClassKind(root.classes(sym))
+      for {
+        kenv <- inferType(tpe0, KindMatch.subKindOf(kind), root)
+        tpeVal = visitType(tpe0, KindMatch.subKindOf(kind), kenv, root)
+        tconstrsVal = Validation.traverse(tconstrs0)(visitTypeConstraint(_, kenv, root))
+        defsVal = Validation.traverse(defs0)(visitDef(_, kenv, root))
+        result <- Validation.sequenceT(tpeVal, tconstrsVal, defsVal)
+        (tpe, tconstrs, defs) = result
+      } yield KindedAst.Instance(doc, mod, sym, tpe, tconstrs, defs, ns, loc)
+  }
+
+  /**
+    * Performs kinding on the given def under the given kind environment.
+    */
+  private def visitDef(def0: ResolvedAst.Def, kenv0: KindEnv, root: ResolvedAst.Root)(implicit flix: Flix): Validation[KindedAst.Def, KindError] = def0 match {
+    case ResolvedAst.Def(sym, spec0, exp0) =>
+      for {
+        specKenv <- getKindEnvFromSpec(spec0, root)
+        kenv <- kenv0 ++ specKenv
```

Comment:
Are kenv0 and specKenv disjoint? My guess is yes.

Reply:
Yes. `kenv0` should only contain tparams from an instance type, which would then not be present in a def's spec.